batch/batch/worker/disk.py

```diff
@@ -1,10 +1,23 @@
 import logging
 
-from hailtop.utils import check_shell_output, LoggingTimer
+from hailtop.utils import check_shell_output, LoggingTimer, sleep_and_backoff
 
 log = logging.getLogger('disk')
 
 
+def retry_n_times(n: int, f, *args, **kwargs):
```

Comment:
This is too broad. Let's specifically look for the rate limit message in the body of the response ("Rate Limit Exceeded"). Google describes the rate limiting like this:

> Currently, projects are limited to the following API rate limit categories. Each limit category is counted separately, so you can achieve the maximum limit in each category simultaneously. Rate limits are enforced and automatically refilled at intervals of every 100 seconds. If you exceed the rate limit for an interval, your request should succeed once you have passed that interval. For example, 20 requests/second would translate to 2000 requests within 100 seconds. That means that if you reach a specific limit anytime within 100 seconds, you need to wait for your quota bucket to refresh to make more requests.

If we get a 403 Rate Limit Exceeded, then we might need to wait as long as 100 seconds before we get a chance to perform our operation.

If a cluster of 100 workers are each trying to mount 32 disks, that's 6400 create and attach operations. We'll fill up the first 2000 operations nearly immediately. Then we wait 100s and another 2000 operations will succeed. Then we wait 100s and another 2000 operations will succeed. Another 100s and the remaining 1400 operations will succeed.

Hmm, what is the ideal outcome? We want to mount disks as fast as possible. Batch Driver also needs to be able to manage instances. In the absence of coordination, exponential backoff hopefully allows everyone to get access. The workers should probably start at a higher delay and back off further than the driver so that the driver is more likely to succeed.

We should integrate this directly into the ComputeClient class. The ComputeClient should probably have a configurable initial delay and max delay. I think the driver is fine with the default max delay of 30s. The workers, on the other hand, should probably be allowed to back off past 30s so that batch driver is more likely to succeed.

Reply:
I asked cotton to give this a think and get back to us.